### PR TITLE
143539467 ne create checkpoint

### PR DESCRIPTION
**Pull request by ski-climb**
@tmikeschu @DavidKnott @ethanbennett @case-eee Ready for review

Basically we're doing 3 things here:
1) rename Location to Checkpoint
2) move UID logic from Lot to Checkpoint 
3) return the new checkpoint object upon successful POST request to '/api/v1/checkpoints' since React front-end will need the ethereum_address to be able to send it to the blockchain

Please give it a good once over.  Looking forward to feedback!  Thanks.

[(Closes) #143539467]

### CODE REVIEW

File: app/models/checkpoint.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        errors.add(:ethereum_address, "must start with '0x'")
+      end
+    end
+
+    def ethereum_address_hex
+      unless ethereum_address[2..-1].chars.all? { |letter| letter =~ /\h/ }
+        errors.add(:ethereum_address, "must be composed of hexidecimal characters only")
+      end
+    end
+
```
Comment (DavidKnott) [inline app/models/checkpoint.rb:46]: Great job changing the address creation to checkpoint!

**Review by DavidKnott — APPROVED** [review]
Great work overall, I love how well tested and broken down the eth address generator is.

File: README.md [hunk 2]
```diff
@@ -7,7 +7,7 @@ you know where you coffee *really* came from? Is your apple truly from a
 sustainable farm? This app addresses these problems by adding visibility to the supply chain process by allowing users to add, update, and
 read transactions on the Ethereum blockchain.
 
-This Rails Api  will provide our [React front end]() with the information it needs (ethereum addresses) to query our  [smart-contract ecosystem]() on the Ethereum blockchain.  It will also store information that is not blockchain pertinent about the goods, transactions, and locations we'll be dealing with such as (name, price, location, &c).
+This Rails Api  will provide our [React front end]() with the information it needs (ethereum addresses) to query our  [smart-contract ecosystem]() on the Ethereum blockchain.  It will also store information that is not blockchain pertinent about the goods, transactions, and checkpoints we'll be dealing with such as (name, price, checkpoint, &c).
```
Comment (tmikeschu) [inline README.md:10]: Great forethought on this 👌

File: app/controllers/api/v1/checkpoints_controller.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
+class Api::V1::CheckpointsController < ApplicationController
+
+  def show
+    checkpoint = Checkpoint.find_by(ethereum_address: params[:id])
```
Comment (tmikeschu) [inline app/controllers/api/v1/checkpoints_controller.rb:4]: Is `params[:id]` based on the route wildcard? I hadn't noticed this mismatch before. @ski-climb

File: app/controllers/api/v1/checkpoints_controller.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      render json: checkpoint
+    else
+      head 404
+    end
+  end
+
+  def create
+    checkpoint = Checkpoint.new(checkpoint_params)
+    if checkpoint.save
+      render json: checkpoint, status: 201
```
Comment (tmikeschu) [inline app/controllers/api/v1/checkpoints_controller.rb:15]: clutch!

File: spec/requests/api/v1/checkpoints_spec.rb [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        country: "Colombia", 
+        zipcode: "050022",
+        lat: 6.27053,
+        lng: -75.57211999999998
+      }
+    }
+
+    response_checkpoint = JSON.parse(response.body)
+
+    expect(response.status).to eq 201
```
Comment (tmikeschu) [inline spec/requests/api/v1/checkpoints_spec.rb:20]: Great documentation 👍

**Review by tmikeschu — APPROVED** [review]
Solid work, thanks for balancing heavy feature improvement and busy work!

**Comment by tmikeschu** [comment]
@ethanbennett you're the merger!

**Review by ethanbennett — APPROVED** [review]
Looks great! I think all the pieces are almost ready to come together.

**Pull request merged**